New module: rds_tag

cahlchang opened the pull request:
I think, RDS module requires only the tagging operation.
It is desirable that the module is loosely coupled.
Like ec2-tag Module.
https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/ec2_tag.py

It is different for RDS and EC2 in the tag rules
I made a new module with the same Playbook interface.

##### SUMMARY

Added module to manipulate tags with RDS of AWS.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
 - rds_tag

##### ANSIBLE VERSION
```
$ ansible --version
ansible 2.4.3.0
  config file = /home/user/git/ansible_repo/ansible.cfg
  configured module search path = ['/home/user/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/pyenv/versions/3.6.3/lib/python3.6/site-packages/ansible
  executable location = /usr/local/pyenv/versions/3.6.3/bin/ansible
  python version = 3.6.3 (default, Dec 13 2017, 03:46:07) [GCC 4.8.2]
```

##### ADDITIONAL INFORMATION
This ticket has also been discussed.
#20395

```
$ ansible-playbook my_rds_tag.yml
```

```yml:my_rds_tag.yml
- hosts: localhost
  become: no
  gather_facts: no
  vars:
    profile: "profile my-profile"
    region: us-east-1a
  tasks:
- name: add tags
  role: rds_tag
    instance_name: "my-rds-db"
    region: "{{ region }}"
    profile: "{{ profile }}"
    state: present
    tags:
      tagkey: "tagvalue"

- name: list tags
  role: rds_tag
    instance_name: "my-rds-db"
    region: "{{ region }}"
    profile: "{{ profile }}"
    state: list
  register: register-tag

- debug:
    msg: "{{ register-tag}}"

- name: remove tags
  role: rds_tag
    instance_name: "my-rds-db"
    region: "{{ region }}"
    profile: "{{ profile }}"
    state: absent
    tags:
      tagkey: "tagvalue"
```

cahlchang committed:
Added module to manipulate tags with RDS of AWS.

julienvey commented on lib/ansible/modules/cloud/amazon/rds_tag.py:
```diff
@@ -0,0 +1,157 @@
+      env: sandbox
+'''
+
+try:
```
This would be handled by the `AnsibleAWSModule`

```python
try:
    import botocore
except ImportError:
    pass  # handled by AnsibleAWSModule
```

julienvey commented on lib/ansible/modules/cloud/amazon/rds_tag.py:
```diff
@@ -0,0 +1,157 @@
+        )
+    )
+
+    module = AnsibleModule(
```
You should use `AnsibleAWSModule`, it has a lot of helpers around boto3

julienvey commented on lib/ansible/modules/cloud/amazon/rds_tag.py:
```diff
@@ -0,0 +1,157 @@
+        argument_spec=argument_spec,
+        supports_check_mode=True)
+
+    if not HAS_BOTO:
```
This will be handled by the `AnsiblleAWSModule`

julienvey commented on lib/ansible/modules/cloud/amazon/rds_tag.py:
```diff
@@ -0,0 +1,157 @@
+
+def list_rds_arn(client, instance_name):
+    if instance_name is None:
+        response_instances = client.describe_db_instances()
```
each call to the AWS API should be wrapped in a `try...except`, such as 
```python
try:
    subnets_by_id = describe_subnets_with_backoff(connection, SubnetIds=subnet_ids, Filters=filters)
except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
    module.fail_json_aws(e, msg="Couldn't find subnet with id %s" % subnet_ids)
```

julienvey reviewed (changes requested):
Would be good to have integration tests (See https://github.com/ansible/community/blob/master/group-aws/integration.md)

Also have a look at other module that make use of boto3 and `AnsibleAWSModule`, as an example (such as https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/ec2_vpc_route_table.py)

cahlchang commented:
Thanks for the review!
I will rectify it immediately.

cahlchang committed:
Fix Metadata

cahlchang committed:
Modify document

cahlchang committed:
Adjust import line to guideline

cahlchang committed:
Use AnsibleAWSModule

cahlchang committed:
Introduce exception handling

cahlchang committed:
Functionalization of implementation

cahlchang committed:
Corrected response message

cahlchang commented:
Fixed the following contents.
* Fix to pass integration tests
* Fix to use AnsibleAWSModule
* Changed to use exception trap
* Changes properly as the code becomes longer

Please point out any improvement points.
Thank you!

cahlchang committed:
Add argument

cahlchang committed:
Fix version

cahlchang committed:
Set default to present.

cahlchang committed:
Compliant with PEP8

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_tag.py:
```diff
@@ -0,0 +1,270 @@
+        argument_spec=argument_spec,
+        supports_check_mode=True)
+
+    region, ec2_url, aws_connect_kwargs = get_aws_connection_info(module,
```
Since you're using AnsibleAWSModule you can remove this line and client = boto3_conn(...) and replace it with `client = module.client('rds')`

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_tag.py:
```diff
@@ -0,0 +1,270 @@
+  returned: always
+  type: string
+  sample: Add tag to arn:aws:rds:eu-west-1:xxxxxxxxxxxx:db:db_name instance
+invocation:
```
You don't need to document the invocation.

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_tag.py:
```diff
@@ -0,0 +1,270 @@
+
+from ansible.module_utils.aws.core import AnsibleAWSModule
+from ansible.module_utils.ec2 import (
+    HAS_BOTO3, ec2_argument_spec, boto3_conn, get_aws_connection_info)
```
You can remove these imports except ec2_argument_spec

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_tag.py:
```diff
@@ -0,0 +1,270 @@
+    list_arn = list_rds_arn(module, client, instance_name)
+    list_tags = list_rds_tags(module, client, list_arn, instance_name)
+
+    if state == 'list':
```
This should go in a facts module instead.

mob-morikawa replied:
I fully agree with the opinion that the function of list should be done in the facts module.

cahlchang replied:
I fully agree with the opinion that the functions of the list should be done in fact modules.

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_tag.py:
```diff
@@ -0,0 +1,270 @@
+
+    if state == 'absent':
+        response = remove_rds_tags(module, client, configured_tags, tags)
+        module.exit_json(changed=True,
```
You could unindent this and remove the one in `if state == 'present'`

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_tag.py:
```diff
@@ -0,0 +1,270 @@
+    if state == 'absent':
+        response = remove_rds_tags(module, client, configured_tags, tags)
+        module.exit_json(changed=True,
+                         **camel_dict_to_snake_dict(response))
```
It would be nice if the current tags were added to the response.

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_tag.py:
```diff
@@ -0,0 +1,270 @@
+    argument_spec = ec2_argument_spec()
+    argument_spec.update(
+        dict(
+            instance_name=dict(type='str'),
```
You can change this to `instance_name=dict(required=True)` and remove the check/failure below if it isn't provided. You don't need to specify `type='str'` because it is the default. I don't think the state 'list' belongs in the module, but for future reference, if you want something to be required for only certain states you could use:
```
module = AnsibleAWSModule(
    argument_spec=argument_spec,
    supports_check_mode=True,
    required_if=[('state', 'present', ('instance_name',)),
                 ('state', 'absent', ('instance_name',))]
)
```

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_tag.py:
```diff
@@ -0,0 +1,270 @@
+    argument_spec.update(
+        dict(
+            instance_name=dict(type='str'),
+            tags=dict(type='dict'),
```
You can add required=True here too and remove the check below.

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_tag.py:
```diff
@@ -0,0 +1,270 @@
+            module.fail_json_aws(e, msg="Couldn't get %s instance's tag"
+                                 % instance_name)
+        tags = {}
+        for datum in response_tags['TagList']:
```
You can reuse boto3_tag_list_to_ansible_dict from ansible.module_utils.ec2 instead of doing this.

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_tag.py:
```diff
@@ -0,0 +1,270 @@
+from ansible.module_utils.ec2 import (
+    HAS_BOTO3, ec2_argument_spec, boto3_conn, get_aws_connection_info)
+from ansible.module_utils.ec2 import (
+    camel_dict_to_snake_dict, snake_dict_to_camel_dict)
```
snake_dict_to_camel_dict can be removed.

s-hertel reviewed (changes requested):
I'm not sure if this module is the right way to go with this. Ideally tagging would be handled in an RDS instance module. There is some work being done for refactoring/splitting the RDS modules that will allow that: #30746

I left some feedback anyway as this would be a more immediate solution.

cahlchang committed:
Delete list status

cahlchang committed:
Delete unnecessary imports

cahlchang committed:
Delete invocation example

cahlchang committed:
list parameter delete

cahlchang committed:
Use AnsibleAWSModule's client

cahlchang committed:
Fix to make tag implementation applicable to module_utility

This commit includes the following changes

* Changed to handle only one instance
* Use the utility to manipulate the tag list
* Fixed to handle only one instance as list operation no longer exists
* Explicit variable name and function name.

cahlchang committed:
Use compare_aws_tags.

cahlchang committed:
Changed to delete only the tag of the intersection set.

cahlchang committed:
Delete unnecessary intents

cahlchang commented:
As you pointed out, we believe that the operation of RDS instances needs to be done within the rds_instance module.
In the rds_instance module, I think that it is desirable to have a tagging function.

However, we also need the rds_tag module to manipulate tags only.
For example, such an example is conceivable.
* An already running RDS instance
* An RDS instance whose tag rule was changed after the operation
* An RDS instance that needs to be marked under certain conditions

In the rds_instance module and the rds_tag module, I think the usage scene is different.

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_tag.py:
```diff
@@ -0,0 +1,223 @@
+        module.exit_json(changed=True,
+                         **camel_dict_to_snake_dict(response))
+
+    response = remove_rds_tags(module, client, tags_configured, tags_operate)
```
Shouldn't this line be indented under `else:` or `if state == 'absent':`?

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_tag.py:
```diff
@@ -0,0 +1,223 @@
+
+    if state == 'present':
+        response = add_rds_tags(module, client, tags_configured, tags_operate)
+        module.exit_json(changed=True,
```
You can remove this `module.exit_json` since the one on line 219 can be used.

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_tag.py:
```diff
@@ -0,0 +1,223 @@
+
+
+def remove_rds_tags(module, client, tags_configured, tags_operate):
+    tags_inter = set(tags_operate.keys()) & set(tags_configured['tags'].keys())
```
compare_aws_tags returns the tags to add and the tags to remove. The tags to remove are in the format of a list of keys. You could call that once in main rather than handle it in each function (you'll need to pass True as the last argument to compare_aws_tags, so it returns the list to remove if state is absent).

s-hertel reviewed (commented):
This module needs integration tests: https://github.com/ansible/community/blob/master/group-aws/integration.md#policy

s-hertel commented:
@cahlchang I think those use cases would still fall under the rds instance module; the module should be idempotent and able to update modifiable attributes (including tags) of the RDS instance using compare_aws_tags. The 'rds_instance' module in the pull request that I linked above does that in the update_rds_tags function: https://github.com/mikedlr/ansible/blob/0873e0e0eb1aa5cde5c422cd6414570b33cb61f8/lib/ansible/modules/cloud/amazon/rds_instance.py#L523. Many modules have a `purge_tags` boolean option. If you simply wanted to add an additional tag you would be able to do something like:
```
rds_instance:
    ...
    purge_tags: no
    tags:
        additional: tag
```
If you wanted to reset the tags to whatever you have in the play you could set purge_tags to true. It seems like the only benefit of this module is that you wouldn't need to pass engine and db_instance_class for state present (https://github.com/mikedlr/ansible/blob/0873e0e0eb1aa5cde5c422cd6414570b33cb61f8/lib/ansible/modules/cloud/amazon/rds_instance.py#L1024). Not sure if that is worth the duplicate logic.

cahlchang commented:
Thank you for the review.

> It seems like the only benefit of this module is that you are not need to pass engine and db_instance_class for state present

I think that's exactly the case.And I think it will be a big feature.By separating the engine and instance type values, Playbook using this module can handle only the information layer of RDS.
This feature works well when you perform operations to set tags of RDS with Ansible.Even if the instance class or engine of RDS is changed, it can be used without changing the setting.It means that RDS's tag-based operation can be performed only with the native Ansible module.If you do the same now, you need to use the shell module frequently.

However, this may not necessarily have to be done with the rds_tag module.I think that it would be good if we had an interface that abstractly treats only tags of RDS.I think that it is also good that there is a state that only the tag is operated on the rds_instance module etc.In that case, 
I think it is important to compare the problem of one module growing and the problem of redundant logic.What seems to be the best is to create the rds_util class and implement the tag operation the same, but first I think that it is sufficient to have similar behavior in the two modules.

How do you think about it?

As for everything else really, it is exactly the case, so if the community needs the rds_tag module, I will start on the fix.

yeukhon commented:
@cahlchang this is a feature I am looking forward to having in Ansible. Ideally, the user interface should just be

```
rds:
  command: tagging
  instance_name: "foo"
  tags:
    tag_1: "value1"
    tag_2: "value2"
  state: present
```

cahlchang commented:
Thank you for your opinion.
I agree.

I think it is also wonderful to have tag operating status in rds module and rds_instance module.
As much as that, I think it is a good thing to be able to manipulate the rds tag with another small module.

I think it is wonderful to be able to define many resource states by Ansible.
And this module should work fine.

cahlchang committed:
fix author

cahlchang committed:
increase required version number.

cahlchang commented:
I confirmed that the function of this PR was satisfied by the rds_instance module.
see
https://github.com/ansible/ansible/blob/devel/test/integration/targets/rds_instance/tasks/test_tags.yml#L84-L93
This PR is already unnecessary, so close it.
I want to express my utmost respect to the creator of rds_instance.
